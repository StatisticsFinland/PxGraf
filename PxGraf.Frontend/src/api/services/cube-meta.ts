/* istanbul ignore file */

import ApiClient from "api/client";
import { useQuery } from "react-query";
import { ICubeMeta } from "types/cubeMeta";
import { defaultQueryOptions } from "utils/ApiHelpers";

/**
 * Interface for cube metadata result. Cube is a multi-dimensional result of Px file data filtered by a query.
 * @property {boolean} isLoading - Flag to indicate if the data is still loading.
 * @property {boolean} isError - Flag to indicate if an error occurred during loading.
 * @property {ICubeMeta} data - The cube meta data as an @see {@link ICubeMeta} object.
 */
export interface ICubeMetaResult {
    isLoading: boolean;
    isError: boolean;
    data: ICubeMeta;
}

const fetchCubeMeta = async(idStack: string[]): Promise<ICubeMeta> => {
    const client = new ApiClient();
    const url = `creation/cube-meta/${idStack.join('/')}`;

    return await client.getAsync(url);
}

export const useCubeMetaQuery = (idStack: string[]): ICubeMetaResult => {
    return useQuery(
        ['cube-meta', idStack],
        () => fetchCubeMeta(idStack),
        defaultQueryOptions
    );
}