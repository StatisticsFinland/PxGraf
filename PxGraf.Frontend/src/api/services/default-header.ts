/* istanbul ignore file */

import ApiClient from "api/client";
import { merge } from "lodash";
import { useQuery } from "react-query";
import { MultiLanguageString } from "types/multiLanguageString";
import { Query } from "types/query";
import { buildTableReference, defaultQueryOptions } from "utils/ApiHelpers";

/**
 * Interface for the chart's header result
 * @property {boolean} isLoading - Flag to indicate if the data is still loading.
 * @property {boolean} isError - Flag to indicate if an error occurred while retrieving the header.
 * @property {MultiLanguageString} data - The header data represented as a multi-language string.
 */
export interface IHeaderResult {
    isLoading: boolean;
    isError: boolean;
    data: MultiLanguageString;
}

const fetchDefaultHeader = async(idStack: string[], query: Query): Promise<MultiLanguageString> => {
    const client = new ApiClient();
    const requestBody = JSON.stringify(
        merge(
            { tableReference: buildTableReference(idStack) },
            { variableQueries: query }
        )
    );
  
    const url = 'creation/default-header';
    return await client.postAsync(url, requestBody);
}

export const useDefaultHeaderQuery = (idStack: string[], query: Query): IHeaderResult => {
    return useQuery(
        ['default-header', ...idStack, query],
        () => fetchDefaultHeader(idStack, query),
        {
            ...defaultQueryOptions,
            enabled: query != null,
        }
    );
}