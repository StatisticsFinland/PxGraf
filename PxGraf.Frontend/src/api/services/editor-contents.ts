/* istanbul ignore file */

import ApiClient from "api/client";
import { useQuery } from "react-query";
import { IEditorContentsResponse } from "types/editorContentsResponse";
import { buildCubeQuery, defaultQueryOptions, useDebounceState } from "utils/ApiHelpers";
import { ICubeQuery, Query } from "types/query";

/**
 * Interface for editor contents result. Editor contents are used by the editor to populate the settings for adjusting the visualization.
 * @property {boolean} isLoading - Flag to indicate if the data is still loading.
 * @property {boolean} isError - Flag to indicate if an error occurred during loading.
 * @property {IEditorContentsResponse} data - The editor contents as an @see {@link IEditorContentsResponse} object.
 */ 
export interface IEditorContentsResult {
    isLoading: boolean;
    isError: boolean;
    data: IEditorContentsResponse;
}

const fetchEditorContents = async (idStack: string[], query: Query, metaEdits: ICubeQuery): Promise<IEditorContentsResponse> => {
    const client = new ApiClient();
    const url = 'creation/editor-contents';
    const requestBody = JSON.stringify(buildCubeQuery(query, metaEdits, idStack));
    return await client.postAsync(url, requestBody);
}

export const useEditorContentsQuery = (idStack: string[], query: Query, cubeQuery: ICubeQuery): IEditorContentsResult => {
    [idStack, query, cubeQuery] = useDebounceState(1000, idStack, query, cubeQuery);
    const queryKey = ['editor-contents', ...idStack, query, cubeQuery];
    return useQuery(
        queryKey,
        () => fetchEditorContents(idStack, query, cubeQuery),
        {
            ...defaultQueryOptions,
            enabled: query != null,
            keepPreviousData: true,
        }
    );
}