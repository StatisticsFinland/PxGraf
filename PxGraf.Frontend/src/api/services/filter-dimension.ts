/* istanbul ignore file */

import ApiClient from "api/client";
import { useQuery } from "react-query";
import { IValueFilter, Query } from "types/query";
import { buildTableReference, defaultQueryOptions } from "utils/ApiHelpers";

/**
 * Interface for variable filter result
 * @property {boolean} isLoading - Flag to indicate if the data is still loading.
 * @property {boolean} isError - Flag to indicate if an error occurred during loading.
 * @property {{ [key: string]: string[] }} data - The variable filter data represented by a dictionary. Key is the variable code and value is a list of strings that represent value codes for that variable.
 */
export interface IFilterVariableResult {
    isLoading: boolean;
    isError: boolean;
    data: {[key: string]: string[]}
}

const fetchResolveVariableFilter = async (idStack: string[], variableFilters: { [key: string]: IValueFilter }): Promise<{ [key: string]: string[] }> => {
    const client = new ApiClient();
    const requestBody = JSON.stringify(
        {
            tableReference: buildTableReference(idStack),
            filters: variableFilters,
        }
    );

    const url = 'creation/filter-dimension'
    return await client.postAsync(url, requestBody);
}

export const useResolveVariableFiltersQuery = (idStack: string[], query: Query): IFilterVariableResult => {
    const varQueries = Object.entries(query ?? {});
    const varFilters =  Object.fromEntries(varQueries.map(([variableCode, variableQuery]) => {
        return [variableCode, variableQuery.valueFilter]
    }));

    return useQuery(
        ['filter-dimension', ...idStack, query],
        () => fetchResolveVariableFilter(idStack, varFilters),
        defaultQueryOptions
    );
}