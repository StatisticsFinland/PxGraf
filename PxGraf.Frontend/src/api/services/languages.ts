/* istanbul ignore file */

import ApiClient from "api/client";
import { useQuery } from "react-query";
import { defaultQueryOptions } from "utils/ApiHelpers";

const fetchLanguages = async (): Promise<string[]> => {
    const client = new ApiClient();
    const url = "creation/languages/";

    return await client.getAsync(url);
};

/**
 * Parses a list of languages into a string for display.
 * @param {string[]} languages List of languages.
 * @returns All listed languages in one string, in upper case.
 */
export const parseLanguageString = (languages: string[]): string => {
    return `(${languages.join(", ").toUpperCase()})`;
}

/**
 * Chooses a listing language for the given listing item, trying to prefer the selected uiLanguage.
 * @param {string} id Id of the database or table.
 * @param {string[]} languages Languages supported by the listing item.
 * @param {string} uiLanguage Currently selected UI language.
 * @returns UI language, if it is supported by the listing item's languages. Otherwise the default language of the item.
 */
export const getDatabaseListingLanguage = (id: string, languages: string[], uiLanguage: string): string => {
    return languages.includes(uiLanguage) ? uiLanguage : languages[0];
}

export const useLanguagesQuery = (): string[] => {
    const queryResult = useQuery(
        ['languages'],
        () => fetchLanguages(),
        defaultQueryOptions
    );
    return queryResult.data;
};