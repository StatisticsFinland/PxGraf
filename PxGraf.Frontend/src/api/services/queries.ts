/* istanbul ignore file */

import ApiClient from "api/client";
import { useMutation, useQuery } from "react-query";
import { ICubeQuery, Query } from "types/query";
import { IVisualizationSettings } from "types/visualizationSettings";
import { buildCubeQuery, defaultQueryOptions } from "utils/ApiHelpers";

/**
 * Response object for saving a query.
 * @property {string} id - The id of the saved query.
 */
export interface ISaveQueryResponse {
    id: string;
}

/**
 * Response object for fetching a saved query.
 * @property query - The query object consisting of the table reference and variable queries @see {@link Query}.
 * @property {IVisualizationSettings} settings - The visualization settings for the query.
 */
export interface IFetchSavedQueryResponse {
    query: {
        tableReference: {
            name: string;
            hierarchy: string[];
        };
        variableQueries: Query;
    } & ICubeQuery;
    settings: IVisualizationSettings;
}

/**
 * Interface for save query result
 * @property {boolean} isLoading - Flag to indicate if the data is still loading.
 * @property {boolean} isError - Flag to indicate if an error occurred during loading.
 * @property {boolean} isSuccess - Flag to indicate if the request was successful.
 * @property {ISaveQueryResponse} data - The save query response data.
 * @property {function} mutate - Function to mutate the state of the query.
 */
export interface ISaveQueryResult {
    isLoading: boolean;
    isError: boolean;
    isSuccess: boolean;
    data: ISaveQueryResponse;
    mutate: (property: boolean) => void;
}

/**
 * Interface for fetch saved query result
 * @property {boolean} isLoading - Flag to indicate if the data is still loading.
 * @property {boolean} isError - Flag to indicate if an error occurred during loading.
 * @property {boolean} isSuccess - Flag to indicate if the request was successful.
 * @property {IFetchSavedQueryResponse} data - The fetch saved query response data.
 */
export interface IFetchSavedQueryResult {
    isLoading: boolean;
    isError: boolean;
    isSuccess: boolean;
    data: IFetchSavedQueryResponse;
}

const sendSaveRequest = async (
    idStack: string[],
    query: Query,
    metaEdits: ICubeQuery,
    selectedVisualization: string,
    visualizationSettings: IVisualizationSettings,
    archive: boolean | void
): Promise<ISaveQueryResponse> => {

    const client = new ApiClient();

    const requestBody = JSON.stringify({
        query: buildCubeQuery(query, metaEdits, idStack),
        settings: {
            ...visualizationSettings,
            selectedVisualization: selectedVisualization,
        }
    });
    const url = archive ? 'sq/archive' : 'sq/save';
    return await client.postAsync(url, requestBody);
}

export const fetchSavedQuery = async (queryId: string): Promise<IFetchSavedQueryResponse> => {
    const client = new ApiClient();
    const url = 'sq/' + queryId;
    const response = await client.getAsync(url);

    // Check for VariableQueries and assign to variableQueries
    if (!response.query?.variableQueries && response.query?.VariableQueries) {
        response.query.variableQueries = response.query.VariableQueries;
    }

    return response as IFetchSavedQueryResponse;
};

export const useFetchSavedQuery = (queryId: string): IFetchSavedQueryResult => {
    return useQuery(
        [queryId],
        () => fetchSavedQuery(queryId),
        defaultQueryOptions
    );
};

export const useSaveMutation = (idStack: string[], query: Query, metaEdits: ICubeQuery, selectedVisualization: string, visualizationSettings: IVisualizationSettings): ISaveQueryResult => {
    return useMutation(
        (archive) => sendSaveRequest(idStack, query, metaEdits, selectedVisualization, visualizationSettings, archive),
        { retry: false }
    );
}