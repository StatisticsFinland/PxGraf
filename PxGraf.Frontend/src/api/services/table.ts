/* istanbul ignore file */

import ApiClient from "api/client";
import { useQuery } from "react-query";
import { defaultQueryOptions } from "utils/ApiHelpers";
import { MultiLanguageString } from "../../types/multiLanguageString";

/**
 * Interface for a table list response.
 * @property {string} id - The id of the table.
 * @property {string} type - The type of the table where 't' is a table and 'l' is a list item.
 * @property {string} updated - The last updated date of the table.
 * @property {MultiLanguageString} text - The description of the table as a multi-language string.
 * @property {string[]} languages - Available languages of the table.
 */
export interface ITableListResponse {
    id: string;
    type?: 't' | 'l';
    updated?: string;
    text: MultiLanguageString;
    languages?: string[];
}

/**
 * Interface for a table result.
 * @property {boolean} isLoading - Flag to indicate if the data is still loading.
 * @property {boolean} isError - Flag to indicate if an error occurred during loading.
 * @property {ITableListResponse[]} data - The table list response data.
 */
export interface ITableResult {
    isLoading: boolean;
    isError: boolean;
    data: ITableListResponse[];
}

const fetchTable = async (idStack: string[], lang: string): Promise<ITableListResponse[]> => {
    const client = new ApiClient();
    const path = idStack.join("/");
    const url = "creation/data-bases/" + path;
    const getParams = {
        'lang': lang
    }

    return await client.getAsync(url, getParams);
};

export const sortTableData = (data: ITableListResponse[], primaryLanguage: string, databaseLanguages: string[]): ITableListResponse[] => {
    return [...data].sort((a, b) => {
        const textA = a.text[primaryLanguage] || a.text[databaseLanguages[0]];
        const textB = b.text[primaryLanguage] || b.text[databaseLanguages[0]];

        return textA.localeCompare(textB, primaryLanguage);
    });
};

export const useTableQuery = (idStack: string[], lang: string): ITableResult => {
    // If trying to fetch a database without a language, return empty
    const fetchFunction = lang ?
        () => fetchTable(idStack, lang) :
        () => Promise.resolve([]);

    return useQuery(
        ['table', ...idStack, lang],
        fetchFunction,
        defaultQueryOptions,
    );
};