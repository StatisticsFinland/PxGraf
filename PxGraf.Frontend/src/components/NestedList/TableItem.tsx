import { Alert, AlertTitle, Divider, ListItem, ListItemButton, ListItemIcon, ListItemText } from "@mui/material";
import FileIcon from '@mui/icons-material/InsertDriveFileOutlined';
import React from 'react';
import { Link } from "react-router-dom";
import { useTranslation } from "react-i18next";
import styled from "styled-components";
import { urls } from "Router";
import { UiLanguageContext } from "contexts/uiLanguageContext";
import { parseLanguageString } from 'utils/ApiHelpers';
import { IDatabaseTable } from 'types/tableListItems';

const StyledListItem = styled(ListItem)`
  background-color: #f8f8f8;
`;

const ErrorAlert = styled(Alert)`
  width: 100%;
`;

/**
 * @property {string[]} currentPath Path to the table in in the hierarchy of the table list.
 * @property {IDatabaseTable} item Contains information about the table.
 * @property {number} depth Current depth in the hierarchy of the table list.
 */
interface ITableItemProps {
    currentPath: string[];
    item: IDatabaseTable;
    depth: number;
}

/**
 * Component used for displaying a table in @see {@link NestedList} component.
 */
export const TableItem: React.FC<ITableItemProps> = ({ currentPath, item, depth }) => {
    const { t } = useTranslation();
    const { language } = React.useContext(UiLanguageContext);
    const displayLanguage = item.languages.includes(language) ? language : item.languages[0];

    return <React.Fragment key={`${item.fileName}-key`}>
        <StyledListItem id={currentPath.join('-')}>
            <ListItemButton sx={{ mr: 3, pl: depth * 4 }} component={Link} to={urls.editor(currentPath)}>
                {item.error ?
                    <ErrorAlert sx={{ pl: depth * 4 }} severity="warning">
                        <AlertTitle>{`${item.name[displayLanguage] ?? item.fileName}`}</AlertTitle>
                        {t("error.contentVariableMissing")}
                    </ErrorAlert>
                    :
                    <React.Fragment>
                        <ListItemIcon sx={{ minWidth: '32px' }}>
                            <FileIcon />
                        </ListItemIcon>
                        <ListItemText primary={`${item.name[displayLanguage]} ${parseLanguageString(item.languages)}`} secondary={t("tableSelect.updated") + ": " + new Date(item.lastUpdated).toLocaleString(language)} />
                    </React.Fragment>
                }
            </ListItemButton>
        </StyledListItem>
        <Divider />
    </React.Fragment>
}