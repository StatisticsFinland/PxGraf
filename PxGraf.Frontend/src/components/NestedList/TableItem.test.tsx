import React from 'react';
import { render } from '@testing-library/react';
import UiLanguageContext from 'contexts/uiLanguageContext';
import '@testing-library/jest-dom';
import { MemoryRouter } from 'react-router-dom';
import { TableItem } from './TableItem';
import { IDatabaseTable } from 'api/services/table';

jest.mock('react-i18next', () => ({
    ...jest.requireActual('react-i18next'),
    useTranslation: () => {
        return {
            t: (str: string) => str,
            i18n: {
                changeLanguage: () => new Promise(() => null),
            },
        };
    },
}));

jest.mock('envVars', () => ({
    PxGrafUrl: 'pxGrafUrl.fi/',
    PublicUrl: 'publicUrl.fi/',
    BasePath: ''
}));

const mockItem: IDatabaseTable = {
    code: 'asd',
    lastUpdated: '2021-10-13T14:53:06',
    name: { 'fi': 'seppo', 'sv': 'seppo-sv' },
    languages: ['fi', 'sv']
}

const mockDepth = 3;
const mockPath = ['foo', 'bar', 'baz'];

const setLanguage = jest.fn();
const language = 'fi';

const setLanguageTab = jest.fn();
const languageTab = 'fi';
const availableUiLanguages = ['fi', 'en', 'sv'];

const uiContentLanguage = "fi";
const setUiContentLanguage = jest.fn();

describe('Rendering test', () => {
    it('renders correctly', () => {
        const { asFragment } = render(
            <MemoryRouter>
                <UiLanguageContext.Provider value={{ language, setLanguage, languageTab, setLanguageTab, availableUiLanguages, uiContentLanguage, setUiContentLanguage }}>
                    <TableItem currentPath={mockPath} item={mockItem} depth={mockDepth} />
                </UiLanguageContext.Provider>
            </MemoryRouter>);
        expect(asFragment()).toMatchSnapshot();
    });
});

describe('Assertion tests', () => {
    it('shows available languages and last updated with table level item names', async () => {
        const { findByText } = render(
            <MemoryRouter>
                <UiLanguageContext.Provider value={{ language, setLanguage, languageTab, setLanguageTab, availableUiLanguages, uiContentLanguage, setUiContentLanguage }}>
                    <TableItem currentPath={mockPath} item={mockItem} depth={mockDepth}  />
                </UiLanguageContext.Provider>
            </MemoryRouter>);

        const name = await findByText("seppo (FI, SV)");
        const lastUpdate = await findByText("tableSelect.updated: 13.10.2021 klo 14.53.06");
        expect(name).toBeInTheDocument();
        expect(lastUpdate).toBeInTheDocument();
    });

    it('renders MUI alert when error property is true', async () => {
        const mockErrorItem: IDatabaseTable = {
            code: 'error',
            lastUpdated: null,
            name: { 'fi': 'error', 'sv': 'error-sv' },
            languages: ['fi', 'sv'],
            error: true
        };

        const { findByRole, findByText } = render(
            <MemoryRouter>
                <UiLanguageContext.Provider value={{ language, setLanguage, languageTab, setLanguageTab, availableUiLanguages, uiContentLanguage, setUiContentLanguage }}>
                    <TableItem currentPath={mockPath} item={mockErrorItem} depth={mockDepth}  />
                </UiLanguageContext.Provider>
            </MemoryRouter>
        );

        const alert = await findByRole('alert');
        const alertHeader = await findByText("error");
        expect(alert).toBeInTheDocument();
        expect(alertHeader).toBeInTheDocument();
    });
});