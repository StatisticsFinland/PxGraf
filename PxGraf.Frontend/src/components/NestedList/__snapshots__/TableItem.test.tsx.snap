// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering test renders correctly 1`] = `
<DocumentFragment>
  <li
    class="MuiListItem-root MuiListItem-gutters MuiListItem-padding sc-gjZUHa iMmsjC css-cfd5wn-MuiListItem-root"
    id="foo-bar-baz"
  >
    <a
      class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-1j5cvue-MuiButtonBase-root-MuiListItemButton-root"
      href="/editor/foo/bar/baz/"
      tabindex="0"
    >
      <div
        class="MuiListItemIcon-root css-sx7q2d-MuiListItemIcon-root"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
          data-testid="InsertDriveFileOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M14 2H6c-1.1 0-1.99.9-1.99 2L4 20c0 1.1.89 2 1.99 2H18c1.1 0 2-.9 2-2V8zM6 20V4h7v5h5v11z"
          />
        </svg>
      </div>
      <div
        class="MuiListItemText-root MuiListItemText-multiline css-14ln1j6-MuiListItemText-root"
      >
        <span
          class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
        >
          seppo (FI, SV)
        </span>
        <p
          class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-1a1whku-MuiTypography-root"
        >
          tableSelect.updated: 13.10.2021 klo 14.53.06
        </p>
      </div>
    </a>
  </li>
  <hr
    class="MuiDivider-root MuiDivider-fullWidth css-1jpc804-MuiDivider-root"
  />
</DocumentFragment>
`;
